### ISSUE TITLE
Request: also provide the insert-citation tool in source mode (it is só useful!)

### ISSUE THREAD

**Issue by florisvdh** [rstudio/rstudio#7876]
Thank you for all the efforts in making RStudio! This feature request has loose relationships with #7276 .

**Question**: Is it planned to provide the insert-citation tool (from visual mode) in _source mode_? That would be a great step IMHO. More experienced Rmd users will most probably prefer to work in source mode, but then would miss out this much relevant tool.

RStudio's insert-citation tool, compared to currently available tools (packages) in R, provides a unique combination of features which are really interesting in a collaborative, heterogeneous environment:

- ability to read both from bibliography files _and_ from Zotero
- ability to read and write to _csl-json, csl-yaml and bib_
- ability to amend existing bibliography files _without touching_ existing content

BTW did you also have a look at the [`rbbt`](https://github.com/paleolimbot/rbbt) package? While it does not combine all these features, it too has a very performant approach to connecting with Zotero; it may be inspiring.

**Comment by ronblum** [comment]
@florisvdh Thank you for the suggestion, and for your kind words! I'll mark this as an enhancement to review, and I'll share your thoughtful sentiment with the rest of the team. :)

**Comment by florisvdh** [comment]
Activity.

**Comment by florisvdh** [comment]
Activity :smile:

**Comment by DominiqueMakowski** [comment]
> Is it planned to provide the insert-citation tool (from visual mode) in source mode? That would be a great step IMHO. More experienced Rmd users will most probably prefer to work in source mode, but then would miss out this much relevant too

Just to say that this would be a really useful feature, the insert citation box is really convenient to link with zotero/crossref and it's quite odd to not be able to access it from source mode (even using the shortcut).

**Comment by Robinlovelace** [comment]
I fully agree this would be useful! As of a few weeks ago I cannot seem to get the citation dialogue to appear *at all* on dailly builds. Have you had similar issues @florisvdh out of interest? Thanks for keeping this issue alive BTW.

**Comment by florisvdh** [comment]
Hi @Robinlovelace, thanks for sharing your finding and thoughts. Since I prefer to use the source mode, I use `rbbt` to insert selected Zotero records in a YAML file, and then mostly make the citation by hand. But having RStudio's tool available in source mode would be a great move forward IMHO.

Your finding about currently non-functional citation dialogue looks as enough reason to make another issue! I didn't test, currently on stable RStudio.

**Comment by ronblum** [comment]
Definitely—@Robinlovelace would you be willing to open a new issue for it? Thanks!

**Comment by ronblum** [comment]
Triage note: Popular request! 😃

**Comment by Robinlovelace** [comment]
> Definitely—@Robinlovelace would you be willing to open a new issue for it? Thanks!

Sure: https://github.com/rstudio/rstudio/issues/11198

**Comment by ronblum** [comment]
Ah, you already did—thank you!

**Comment by mikebessuille** [comment]
Per JJ: right now, the component that does this is a little too coupled to the Visual editor; work to decouple that will take some time, at which point we can use it to support the source editor

**Comment by dmi3kno** [comment]
I would like to reiterate this request. Several posts appeared in the [Posit Community](https://community.rstudio.com/t/insert-citation-dialog-in-default-plain-source-markdown-editor-and-not-visual-mode/98466), but I am glad I discovered this! Please keep up the good work. Insert citation dialogue is a blessing we would like to use more!

**Comment by Steinthal** [comment]
Check the Rstudio Addin [citr](https://github.com/crsh/citr) to insert Markdown Citations in the source view.

**Comment by florisvdh** [comment]
Thanks @Steinthal, I explored this a few years ago, but AFAIK **citr** doesn't support the CSL (yaml or json) bibliographic file format, while CSL is Pandoc's native format and hence provides best support for all R Markdown output formats. Also, **citr** struggled with larger Zotero databases, contrary to RStudio VME and the **rbbt** package.

Regarding the recommended CSL format, note that using a `.bib` file in between Zotero and Pandoc comes down to a lossy back-and-forth-translation process from and to the CSL format (csl -> bib -> csl). For more information, [see BBT](https://retorque.re/zotero-better-bibtex/exporting/pandoc/#use-csl-not-bibtex-with-pandoc).

**Comment by jameswoodgeog** [comment]
Hi - I will reiterate appreciation of the work that goes into RStudio. This function would be great. Unless there is a way of adding a citation to the .bib file without having to go to Visual editor. I would love this function in Source too! My reason (and when I get time, I will raise an issue), is that whenever I go to visual mode to add a reference (so it gets added to the .bib file) while working on a quarto document, my indenting breaks when I go back to Source, and this is unimaginably painful for a novice.

Thanks again!

**Comment by jthomasmock** [comment]
Thanks for the interest! We're making a lot of improvements to the Visual Editor for round-tripping between visual and source mode.

**Comment by LuisLauM** [comment]
Please, do not forget this issue. It would be great to have the ability of insert citations (and References) from Zotero to Quarto in the source mode (and not only in the visual mode).

**Comment by jthomasmock** [comment]
Thanks for your interest! We've reviewed how this _could_ work, but it's currently highly reliant on the Visual Editor tool itself, as it will use a different state context, data, etc. We're unlikely to untangle this in the short term for source mode, but will keep this issue open for the future.
